Issue 44632: Publish study wizard support for selecting user-defined Queries

cnathe opened the pull request:
#### Rationale
https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=44632

Study publish wizard exports & imports every archivable object we have... except queries. That means, for example, that publishing the LabKey demo study results in broken R charts (because they depend on queries). As far as we can tell, the intent was to show queries on their own panel (allowing selecting subsets to publish), but it was never actually wired up. This PR does just that.

#### Related Pull Requests
* https://github.com/LabKey/platform/pull/3034
* https://github.com/LabKey/testAutomation/pull/1004

#### Changes
* update wizard to add a new panel that shows a grid of the user-defined queries for all schemas in the given container
* GetSchemaQueryTreeAction updates to get all user defined queries for a container
* pipe through query key selections to CreateChildStudyPipelineJob and FolderExportContext to be used in QueryWriter
* handle query key selections in study.StudySnapshot definitions
* update the set of Grid Views for selection based on the selected query keys (in addition to the already supported list and dataset selection filtering)

cnathe committed:
Issue 44632: Publish study wizard support for selecting user-defined Queries
* update wizard to add a new panel that shows a grid of the user-defined queries for all schemas in the given container
* GetSchemaQueryTreeAction updates to get all user defined queries for a container
* pipe through query key selections to CreateChildStudyPipelineJob and FolderExportContext to be used in QueryWriter
* handle query key selections in study.StudySnapshot definitions

cnathe committed:
Update the set of Grid Views for selection based on the selected query keys (in addition to the already supported list and dataset selection filtering)

cnathe committed:
StudyPublishTest updates for new Queries panel

cnathe committed:
refactor some study test helpers for advancing through the publish wizard
- for the common cases of changing nothing or selecting all in the grid

cnathe committed:
Merge branch 'develop' into fb_issue44632PublishQueries

cnathe committed:
Merge branch 'develop' into fb_issue44632PublishQueries

cnathe committed:
SharedStudyTest fix for participants panel title check

cnathe committed:
StudyPublishTest fix for study objects and folder objects select all

cnathe committed:
missed from prev commit

cnathe committed:
add null check for a study with no datasets

cnathe committed:
Fix for StudyPublishTest participant panel label check

cnathe committed:
StudyPublishQueriesTest - new test for queries selection in publish study wizard

labkey-tchad commented on query/src/org/labkey/query/QueryWriter.java:
```diff
@@ -68,6 +70,13 @@ public void write(Container object, ImportContext<FolderDocument.Folder> ctx, Vi
         List<QueryDefinition> queries = new ArrayList<>(QueryServiceImpl.get().getQueryDefsAndMetadataOverrides(ctx.getUser(), c));
         FileNameUniquifier fileNameUniquifier = new FileNameUniquifier();
 
+        if (ctx.getClass().equals(FolderExportContext.class))
+        {
+            Set<String> queryKeysToExport = ((FolderExportContext)ctx).getQueryKeys();
+            if (queryKeysToExport != null)
+                queries.removeIf(queryDef -> !queryKeysToExport.contains(queryDef.getQueryKey()));
+        }
```
Why exclude `FolderExportContext` subclasses? Is there a chance they would want this behavior?

```suggestion
        if (ctx instanceof FolderExportContext feCtx)
        {
            Set<String> queryKeysToExport = feCtx.getQueryKeys();
            if (queryKeysToExport != null)
                queries.removeIf(queryDef -> !queryKeysToExport.contains(queryDef.getQueryKey()));
        }

```

cnathe replied:
Good point. I was copying over logic from the ReportWriter. I've updated both to your suggestion. Thanks.

cnathe committed:
Switch StudyPublishQueriesTest to use wiki schema instead of assay.General

labkey-tchad commented on study/test/src/org/labkey/test/tests/study/StudyPublishQueriesTest.java:
```diff
@@ -0,0 +1,224 @@
+
+        // need at least one visit for the publish study wizard
+        _studyHelper.goToManageVisits().goToImportVisitMap();
+        setFormElement(Locator.name("content"),
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" +
+                "<visitMap xmlns=\"http://labkey.org/study/xml\">\n" +
+                "  <visit label=\"Test Visit1\" typeCode=\"X\" sequenceNum=\"1.0\" maxSequenceNum=\"1.0\"/>\n" +
+                "  <visit label=\"Test Visit2\" typeCode=\"X\" sequenceNum=\"2.0\" maxSequenceNum=\"2.0\"/>\n" +
+                "  <visit label=\"Test Visit3\" typeCode=\"X\" sequenceNum=\"3.0\" maxSequenceNum=\"3.0\"/>\n" +
+                "</visitMap>"
+        );
```
Java 16 text blocks!!
```suggestion
        setFormElement(Locator.name("content"),
        """
        <?xml version=\"1.0\" encoding=\"UTF-8\"?>
          <visitMap xmlns=\"http://labkey.org/study/xml\">
            <visit label=\"Test Visit1\" typeCode=\"X\" sequenceNum=\"1.0\" maxSequenceNum=\"1.0\"/>
            <visit label=\"Test Visit2\" typeCode=\"X\" sequenceNum=\"2.0\" maxSequenceNum=\"2.0\"/>
            <visit label=\"Test Visit3\" typeCode=\"X\" sequenceNum=\"3.0\" maxSequenceNum=\"3.0\"/>
          </visitMap>
        """
        );
```

cnathe replied:
Cool! Done.

labkey-tchad commented on study/test/src/org/labkey/test/tests/study/StudyPublishQueriesTest.java:
```diff
@@ -0,0 +1,224 @@
+            createQuery(getProjectName(), USER_DEFINED_QUERY, schemaName, "SELECT 'Test from " + schemaName + " schema'", null, false);
+    }
+
+    @Override // TODO delete me
+    protected void doCleanup(boolean afterTest)
+    {
+        super.doCleanup(afterTest);
+    }
```
Delete me!

cnathe replied:
Thanks, done.

cnathe committed:
Merge branch 'develop' into fb_issue44632PublishQueries

cnathe committed:
Code review feedback
- StudyHelper.participantList()
- advanceThroughPublishStudyWizard() overload which takes single panel

cnathe committed:
CR feedback to use instanceof FolderExportContext check
